pkg/network/master/master.go

```diff
@@ -156,6 +156,13 @@ func Start(networkConfig osconfigapi.MasterNetworkConfig, networkClient networkc
 		return err
 	}
 
+	// FIXME: this is required to register informers for the types we care about to ensure the informers are started.
+	// FIXME: restructure this controller to add event handlers in Start() before returning, instead of inside startSubSystems.
```

Comment:
This was caused by https://github.com/openshift/origin/pull/18911 where we moved some heavy lifting from SDN master controller to process asynchronously (mainly to wait for informers to be synced and use cached values were ever is appropriate).

Looking at other openshift controllers, Additional queue is maintained for each resource (apart from the queue used in shared informer). I'm assuming this was done mainly to process multiple items in parallel by the worker threads(?) (example: https://github.com/openshift/origin/blob/master/pkg/quota/controller/clusterquotamapping/clusterquotamapping.go#L169).

How is this case handled:
time t1 : Namespace n1, Set cpu quota to 2 cores                        // First item in the queue
time t2 (t1 + delta): Namespace n1, Set cpu quota to 4 cores      // Second item in the queue

Worker w1 picked first item, and worker w2 picked second item
time t3: w2 sets cpu quota to 4 cores on namespace n1
time t4: w1 sets cpu quota to 2 cores on namespace n1 which is not desired.
How are the workers ensuring the correct order for the *same* namespace?

If we don't need to process items in parallel, can we live with this workaround (calling `<resourceInformer>.Informer().GetController()`)?

@liggitt @deads2k

Reply:
> If we don't need to process items in parallel, can we live with this workaround (calling <resourceInformer>.Informer().GetController())?


No, this isn't a long term workaround. You want a workqueue to collapse multiple events for the same object as well even if you don't do parallel processing. Also, adding handlers after start disrupts existing listeners (we have to stop the world and replay current state as synthetic events for the late handler)

Reply:
Cluster quota isn't the best controller to look at as a simple watch/process example… it is maintaining a matrix of namespaces/quotas/quotaed objects